xiilib/_gunicorn/wsgi_app.py

```diff
@@ -0,0 +1,131 @@
+        env = {f"{prefix}{k.upper()}": self._encode_env(v) for k, v in config.items()}
+        secret_key_env = f"{prefix}SECRET_KEY"
+        if secret_key_env not in env:
+            env[secret_key_env] = self._charm_state.secret_key
+        for proxy_variable in ("http_proxy", "https_proxy", "no_proxy"):
+            proxy_value = getattr(self._charm_state.proxy, proxy_variable)
+            if proxy_value:
+                env[proxy_variable] = str(proxy_value)
+                env[proxy_variable.upper()] = str(proxy_value)
+        env.update(self._charm_state.database_uris)
+        return env
+
+    def _wsgi_layer(self) -> ops.pebble.LayerDict:
+        """Generate the pebble layer definition for WSGI application.
+
+        Returns:
+            The pebble layer definition for WSGI application.
+        """
+        original_services_file = self._charm_state.state_dir / "original-services.json"
+        if self._container.exists(original_services_file):
+            services = json.loads(self._container.pull(original_services_file).read())
+        else:
+            plan = self._container.get_plan()
+            services = {k: v.to_dict() for k, v in plan.services.items()}
+            self._container.push(original_services_file, json.dumps(services), make_dirs=True)
+
+        services[self._charm_state.service_name]["override"] = "replace"
+        services[self._charm_state.service_name]["environment"] = self._wsgi_environment()
+
+        return ops.pebble.LayerDict(services=services)
+
+    def restart(self) -> None:
+        """Restart or start the WSGI service if not started with the latest configuration."""
+        if not self._container.can_connect():
+            logger.info(
+                "pebble client in the %s container is not ready", self._charm_state.framework
+            )
+            return
+        if not self._charm_state.is_secret_storage_ready:
+            logger.info("secret storage is not initialized")
+            return
+        self._container.add_layer("charm", self._wsgi_layer(), combine=True)
+        service_name = self._charm_state.service_name
+        is_webserver_running = self._container.get_service(service_name).is_running()
+        command = self._wsgi_layer()["services"][self._charm_state.framework]["command"]
+        self._webserver.update_config(
+            environment=self._wsgi_environment(),
+            is_webserver_running=is_webserver_running,
+            command=command,
+        )
+        migration_command = None
+        app_dir = self._charm_state.app_dir
+        if self._container.exists(app_dir / "migrate"):
+            migration_command = [str((app_dir / "migrate").absolute())]
+        if self._container.exists(app_dir / "migrate.sh"):
+            migration_command = ["bash", "-eo", "pipefail", "migrate.sh"]
+        if self._container.exists(app_dir / "migrate.py"):
+            migration_command = ["python", "migrate.py"]
```

Comment:
Shouldn't we use a `match` `case` statement instead ?

Reply:
Yes, but I think it's only available in Python 3.10, which is not the default Python version in ubuntu:22.04

Reply:
Are you sure? I think ubuntu 22.04 ships with 3.10

Reply:
Sorry I mean ubuntu:20.04, do we need to support ubuntu:20.04?

Reply:
Since we're on K8S, I think we should drive the users to that latest LTS OS

Reply:
The OS is decided by `rockcraft.yaml` `base`, which means user can control that (we can reject old LTS in the extension though)

Reply:
yeh thats fine, we don't need to support 20.04

Reply:
Well as long as the default behavior of our rockcraft extension doesn't cause compatibility issue we're fine.
The are many way that users could use with customisation to break the charm / rock file, we can't protect against all them

Reply:
okay, I will remove support for 20.04 in charmcraft and rockcraft extension